Condition support in the library

neithernut opened the pull request.

neithernut committed:
Move trait Condition to new module

We'll introduce new stuff closely related to conditions.

neithernut committed:
Add match operators

neithernut committed:
Add ConditionFactory trait

neithernut committed:
Add error kind for condition parsing errors

neithernut committed:
Add function for parsing bits of a condition

neithernut committed:
Use new parsing utilities for conditions in YAML parser

neithernut committed:
Add new condition factory for test conditions

neithernut committed:
Update documentation of the Conditino trait

... in order to reflect recent changes to the specification.

neithernut committed:
Add some initial tests for condition parsing

neithernut commented:
I would like to make actual parsing (as from strings) available in the `ConditionFactory` trait as a predefined function. It would return a result wrapping a `Condition`. This would be more ergonomic than calling `parse_condition()` and then feeding the result into `ConditionFactory::make_condition()`. However, I'm not sure how to resolve the conflict in the error types and would appreciate some feedback on this.

I'm currently aware of the following options:

 * Just force the implementer of `ConditionFactory::make_condition` to use our error type, maybe add a `String` to `ConditionParseError`.
 * Wrap the user's error type for parsing directly as a member of `ConditionParseError`. This would, introduce all the shenanigans related to dependent error types for _just_ condition parsing.
 * Wrap the user's error type for parsing in a `Box<Error>`. Well, boxes...
 * Use something like [error-chain](https://crates.io/crates/error-chain). Again, we would carry all that load only for parsing.

Any ideas? @matthiasbeyer?

matthiasbeyer commented on src/condition.rs:
```diff
@@ -0,0 +1,195 @@
+///
+pub trait ConditionFactory<C, E>
+    where C: Condition + Sized,
+          E: EError
```
This is a output type. You should make The output type an associated type of the trait, like so:

```
pub trait Too {
    type Par;
}
```

With an optional constraint for the error type on `std::error::Error` (but I'd suggest not to because it is not necessary afaics).

neithernut replied:
Yep. In fact, I planned doing this, already.

matthiasbeyer commented:
Maybe I'm misunderstanding something... To late to think right now. As far as I can see (on mobile atm), you would only need to create a error variant with an user error type wrapped, if you're going to apply my suggestion from above,... I'll have a closer look tomorrow though...

matthiasbeyer commented:
Feel free to apply [this patch](https://github.com/matthiasbeyer/issue-states/commit/1c430d9907310821bcebe4fe9bc265abe1054ec3.patch) to make the error type a dependent type with a conversion constraint. This way you can provide a default implementation in the trait without having implementing code forced to use your own error type.

neithernut commented:
Great idea. Will do as soon as I'm back from work.

matthiasbeyer commented:
You can also ff-merge from my [branch](https://github.com/matthiasbeyer/issue-states/tree/pr-17-patches) if you want to do this.

neithernut committed:
Make TestCondParseError usable for use in Conditions

For this, we have to implement `From<error::Error>`. In addition, we
also implement `Default` in order to ease construction.

neithernut committed:
Add direct parsing capabilities to ConditionFactory trait

neithernut commented:
@matthiasbeyer thanks for the patch. I had to change a few things for it to actually work, though. I'll squash it into the commit introducing the `ConditionFactory` trait at some point, if you are ok with it.
